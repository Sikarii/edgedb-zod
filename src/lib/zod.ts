import { $ } from "edgedb";

const zodType = (name: string, ...args: unknown[]) => `${name}(${args})`;

const zodEnum = (values: unknown[]) => {
  const str = values.join(", ");
  return zodType("enum", `[${str}]`);
};

export const scalarToZod = (type: $.introspect.ScalarType) => {
  if (type.enum_values) {
    const values = type.enum_values.map((v) => `"${v}"`);

    return [
      zodEnum(values),
    ];
  }

  switch (type.name) {
    case "std::str":
      return [
        zodType("string"),
      ];

    case "std::bool":
      return [
        zodType("boolean"),
      ];

    case "std::json":
      return [
        zodType("unknown"),
      ];

    case "std::uuid":
      return [
        zodType("string"),
        zodType("uuid"),
      ];

    case "std::int16":
      return [
        zodType("number"),
        zodType("int"),
        zodType("min", 0),
        zodType("max", 65535),
      ];

    case "std::int32":
      return [
        zodType("number"),
        zodType("int"),
        zodType("min", 0),
        zodType("max", 2147483647),
      ];

    case "std::int64":
      return [
        zodType("number"),
        zodType("int"),
        zodType("min", 0),
      ];

    case "std::bigint":
      return [
        zodType("bigint"),
      ];

    case "std::datetime":
      return [
        zodType("date"),
      ];

    case "std::float32":
      return [
        zodType("number"),
        zodType("min", 0),
        zodType("max", 3.40282347E+38),
      ];

    case "std::float64":
      return [
        zodType("number"),
        zodType("min", 0),
        zodType("max", 1.7976931348623157E+308),
      ];

    // This is not a type in EdgeDB, the JS driver has this
    case "std::number":
      return [
        zodType("number"),
      ];

    default:
      // TODO: Null might make more sense
      return [
        zodType("never"),
      ];
  }
};
