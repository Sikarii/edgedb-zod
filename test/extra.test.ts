import { z } from "zod";
import { it, describe, expect } from "vitest";

const extra = import("../dbschema/edgedb-zod/modules/extra");

describe("extra::MyEnum", async () => {
  const schema = (await extra).MyEnumSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodEnum);
  });

  it("has the expected values", () => {
    expect(schema._def.values).toMatchSnapshot();
  });
});

describe("extra::CreateMyType", async () => {
  const schema = (await extra).CreateMyTypeSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodObject);
  });

  it("has the expected shape", () => {
    expect(schema.shape).toMatchSnapshot();
  });
});

describe("extra::UpdateMyType", async () => {
  const schema = (await extra).UpdateMyTypeSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodObject);
  });

  it("has the expected shape", () => {
    expect(schema.shape).toMatchSnapshot();
  });
});
