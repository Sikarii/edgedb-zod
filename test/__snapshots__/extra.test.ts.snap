// Vitest Snapshot v1

exports[`extra::CreateMyType > has the expected shape 1`] = `
{
  "createdAt": ZodOptional {
    "_def": {
      "innerType": ZodString {
        "_def": {
          "checks": [
            {
              "kind": "regex",
              "regex": /\\^\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}\\(T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\(\\\\\\.\\\\d\\{3\\}\\)\\?\\)\\?Z\\?\\$/,
            },
          ],
          "typeName": "ZodString",
        },
        "_regex": [Function],
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nonempty": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "trim": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "enumTest": ZodOptional {
    "_def": {
      "innerType": ZodEnum {
        "_def": {
          "typeName": "ZodEnum",
          "values": [
            "1",
            "2",
            "3",
            "4",
            "5",
          ],
        },
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "jsonData": ZodOptional {
    "_def": {
      "innerType": ZodUnknown {
        "_def": {
          "typeName": "ZodUnknown",
        },
        "_unknown": true,
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "updatedAt": ZodOptional {
    "_def": {
      "innerType": ZodString {
        "_def": {
          "checks": [
            {
              "kind": "regex",
              "regex": /\\^\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}\\(T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\(\\\\\\.\\\\d\\{3\\}\\)\\?\\)\\?Z\\?\\$/,
            },
          ],
          "typeName": "ZodString",
        },
        "_regex": [Function],
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nonempty": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "trim": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
}
`;

exports[`extra::MyEnum > has the expected values 1`] = `
[
  "1",
  "2",
  "3",
  "4",
  "5",
]
`;

exports[`extra::UpdateMyType > has the expected shape 1`] = `
{
  "enumTest": ZodOptional {
    "_def": {
      "innerType": ZodEnum {
        "_def": {
          "typeName": "ZodEnum",
          "values": [
            "1",
            "2",
            "3",
            "4",
            "5",
          ],
        },
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "jsonData": ZodOptional {
    "_def": {
      "innerType": ZodUnknown {
        "_def": {
          "typeName": "ZodUnknown",
        },
        "_unknown": true,
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "updatedAt": ZodOptional {
    "_def": {
      "innerType": ZodString {
        "_def": {
          "checks": [
            {
              "kind": "regex",
              "regex": /\\^\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}\\(T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\(\\\\\\.\\\\d\\{3\\}\\)\\?\\)\\?Z\\?\\$/,
            },
          ],
          "typeName": "ZodString",
        },
        "_regex": [Function],
        "and": [Function],
        "array": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nonempty": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "promise": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "trim": [Function],
      },
      "typeName": "ZodOptional",
    },
    "and": [Function],
    "array": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "promise": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
}
`;
